packages/from-hardhat/src/api.ts

```diff
@@ -0,0 +1,142 @@
+import { promises as fs } from "fs";
+import semver from "semver";
+
+import type * as Hardhat from "hardhat/types";
+import type * as Common from "@truffle/compile-common";
+import type TruffleConfig from "@truffle/config";
+
+import { supportedHardhatVersionRange } from "./constants";
+import * as Compilation from "./compilation";
+import * as Config from "./config";
+
+import {
+  checkHardhat,
+  askHardhatConsole,
+  askHardhatVersion
+} from "./ask-hardhat";
+import { EnvironmentOptions } from "./options";
+
+/**
+ * Checks for the existence of a Hardhat project configuration and asserts
+ * that the local installed version of Hardhat matches this package's
+ * supported version range.
+ *
+ * @param options to control process environment (e.g. working directory)
+ * @return Promise<void> when expectation holds
+ * @throws NotHardhatError when not in a Hardhat project directory
+ * @throws IncompatibleHardhatError if Hardhat has unsupported version
+ */
+export const expectHardhat = async (
+  options?: EnvironmentOptions
+): Promise<void> => {
+  const isHardhat = await checkHardhat(options);
+
+  if (!isHardhat) {
+    throw new NotHardhatError();
+  }
+
+  const hardhatVersion = await askHardhatVersion(options);
+
+  if (!semver.satisfies(hardhatVersion, supportedHardhatVersionRange)) {
+    throw new IncompatibleHardhatError(hardhatVersion);
+  }
+};
+
+/**
+ * Thrown when no Hardhat project is found
+ */
+export class NotHardhatError extends Error {
```

Comment:
Oh yeah I meant to note that I extended `Error` with the plan that it will be later chained via your #5367 

Seems better to do that vs. import @truffle/error here.

Reply:
Wait, error chaining requires that you extend @truffle/error?

I thought the idea was "throw whatever error you like and @truffle/core will decide if it should be wrapped as TruffleError"

Reply:
FWIW, extending from `TruffleError` makes my work in #5367 substantially easier. This is because I've overridden the `stack` property on `TruffleError` in that PR to spit out stack traces for the underlying causes in addition to the top-level one, as personally I find this to be the most useful reason to chain errors. Unfortunately we don't get that unless we inherit from `TruffleError`, as the standard `stack` property doesn't include the traces from the underlying causes.

That said, if you'd rather not do that now, I'll be happy to rebase on `develop` when this merges and I can handle it then.

Reply:
> Wait, error chaining requires that you extend @truffle/error?
> 
> I thought the idea was "throw whatever error you like and @truffle/core will decide if it should be wrapped as TruffleError"... like, that neatly makes error display a concern of @truffle/core, no?

Nah, that's a concern that's separate from error chaining. Error chaining allows you to catch one error, throw another, but retain a reference to the error that was the original underlying cause of the one that you're throwing. It makes wrapped errors much easier to source back to their root cause.

Reply:
So like

```
} catch(err) {
  throw new WrappedFooError("Foo failed real bad, guys", { cause: err });
}
```

Reply:
I'll leave the error class hierarchy stuff for you with #5367 ;)